docs/configs.rst
.. _configuration:

Configs
#######

Lowball configs come in the form of a YAML or JSON file. There are six recognized top-level configuration sections. Each
section can contain the configuration object for that section or they can point to a json/yaml file containing the
configuration data for that section

`meta`
  Metadata about the application such as description, tags, etc.

`authentication`
  Configs used for the governance of tokens such as max lifetime of a token.

`application`
  Configurations used by the service itself.

`auth_provider`
  All data surrounding the chosen Authentication Provider.

`auth_db`
  All data surrounding the chosen Authentication Database.

`logging`
  All data surrounding the chosen log provider and format.

All configuration values native to lowball have default values and thus have no requirements.

Upon initialization, all configuration values are mapped to the app class variable `lowball_config`, which means that it
can be accessed in the following manner:

**In the Application Itself**

.. code-block:: python

    self.lowball_config


**In a View Function**

.. code-block:: python

    from flask import current_app


    @app.route("/", methods=["GET"])
    def view_func():
        app_config = current_app.lowball_config


Meta Config
************

The `meta` config section is used to define attributes that describe the application. The primary purpose of this data
is exposing it on the `/builtin/status` endpoint. However, like all config items it is made available for use anywhere
in the application. There is one reserved field used by the operation of the application and four reserved fields used
by the status route.

`base_route`
  If provided will be prepended to every route. For example if your have `localhost:5000/hello` and then define a base
  route of `/app` then the full url in operation would be `localhost:5000/app/hello`. This functionality can be useful
  when running the services behind an API Gateway, ingress controller, etc.

`name`
  Simple name of the application.

`description`
  Simple description of your application.

`tags`
  Used to label / tag your application instance.

**Meta Example**

.. code-block:: yaml

    meta:
      name: THE_APP
      base_route: /app
      description: "my super awesome application"
      tags:
        - experimental
        - awesome

Authentication Config
*********************

The `authentication` config used for the governance of tokens in the system. There are three available settings:

`token_secret`
  The token secret is the string used to encrypt / decrypt the JWTs (tokens). It is highly encouraged to set this to a
  secure string of your choosing for anything other than local development. If it is not specified in the config, it
  will default to `CHANGE_ME`.

`default_token_life`
  Specifies the lifetime (in seconds) that a generated token is to be valid, if it is not specified by the requester. If
  a configuration value is not supplied, it will default to 3600 (1 hour).

`max_token_life`
  Specifies the maximum lifetime (in seconds) that can be granted to a token. If a configuration value is not specified
  it will default to 2592000 (~30 days).


.. code-block:: yaml

    authentication:
      default_token_life: 3600
      max_token_life: 7200
      token_secret: "supersecrettokensecret"


Application Config
******************

The `application` config is the place where configs that are specific to the application are meant to be stored. This
could be anything from URLs for external APIs that the service is meant to make requests to, to usernames for databases
that the service pulls data from. Given the nature of this type of data, there is no enforcement on the data that
is contained in this section of the config. All that is necessary is that it be in an object format in the config, so
that it can be read in as a python `dict` object.


Authentication Provider Config
******************************

Configuration values associated with the chosen Authentication Provider. These values are defined by the implementation
of the given provider. See :ref:`available-providers` for further documentation.

Auth Database Config
********************

Configuration values associated with the chosen Authentication Database. These values defined by the implementation of
the given database. See :ref:`available-databases` for further documentation.

Logging Config
**************

Configuration values associated with the chosen Logging Provider. See :ref:`logging` for further documentation.

Reading in Configs
*******************

The configs that lowball uses can be read in using two methods:
1. Directly from a python `dict` using the `config_from_object` builtin method
2. From a JSON or YAML file using the `config_from_file` builtin method

**From an object**

.. code-block:: python

    from lowball import config_from_object

    config = {
       "meta": {
          "name": "APP",
          "base_route": "/app",
          "description": "example to show config reading methods"
       },
       "authentication": {
          "max_token_life": 7200,
          "default_token_life": 3600,
          "token_secret": "supersecrettokensecret"
       },
       "application": {
          "username": "user_of_import"
       },
       "auth_provider": {
          ...
       },
       "auth_db": {
          ...
       },
       "logging": {
          ...
       }
    }

    config_object = config_from_object(config)


**From a JSON File**

`config.json` could look something like this:

.. code-block:: json

    {
       "meta": {
          "name": "APP",
          "base_route": "/app",
          "description": "example to show config reading methods"
       },
       "authentication": {
          "max_token_life": 7200,
          "default_token_life": 3600,
          "token_secret": "supersecrettokensecret"
       },
       "application": {
          "username": "user_of_import"
       },
       "auth_provider": {

       },
       "auth_db": {

       },
       "logging": {

       }
    }


We would read it in like this:

.. code-block:: python

    from lowball import config_from_file

    config_object = config_from_file("./config.json")


**From a YAML File**

`config.yaml` could look something like this:

.. code-block:: yaml

    meta:
      name: APP
      base_route: /app
      description: "description of application goes here"
    authentication:
      default_token_life: 3600
      max_token_life: 7200
      token_secret: "supersecrettokensecret"
    application:
      username: user_of_import
    auth_provider:
      ...
    auth_db:
      ...
    logging:
      ...

or this if using sub files:

.. code-block:: yaml

    meta: /path/to/meta.yaml
    authentication:
      default_token_life: 3600
      max_token_life: 7200
      token_secret: "supersecrettokensecret"
    application: /path/to/app.yaml
    auth_provider:
      ...
    auth_db:
      ...
    logging:
      ...

We would read it in like this:

.. code-block:: python

    from lowball import config_from_file

    config_object = config_from_file("./config.yaml")


lowball/models/config_models/config.py
from pathlib import Path

from ruamel.yaml import YAML

class MetaConfig:
    """Object representation of the Meta config section of a Lowball config.

    Defines metadata about the Lowball application.

    :param name: The name of the Lowball application that is using this config defaults to "lowball"
    :type name: str
    :param base_route: The base route of the Lowball application. This route will
        be at the base of every route in the application. defaults to the root "/"
    :type base_route: str
    :param description: A description of the Lowball application that is using this
        config.
    :type description: str
    :param tags: Tags associated with this Lowball application
    :type tags: list
    """

    # TODO: Implement the ability for someone to add arbitrary metadata

    def __init__(self, name="lowball", base_route="/", description="", tags=None, **kwargs):
        self.name = name
        self.base_route = base_route
        self.description = description
        self.tags = tags

    @property
    def name(self):
        return self._name

    @name.setter
    def name(self, value):
        if not isinstance(value, str):
            raise TypeError("name must be a string")
        self._name = value

    @property
    def base_route(self):
        return self._base_route

    @base_route.setter
    def base_route(self, value):
        if not isinstance(value, str):
            raise TypeError("base_route must be a string")
        if not value.startswith("/"):
            raise ValueError("base_route must start with a forward slash (/)")
        self._base_route = value

    @property
    def description(self):
        return self._description

    @description.setter
    def description(self, value):
        if not isinstance(value, str):
            raise TypeError("description must be a string")
        self._description = value

    @property
    def tags(self):
        return self._tags

    @tags.setter
    def tags(self, value):
        if value is None:
            value = []
        if not isinstance(value, list):
            raise TypeError("tags must be a list of strings")
        if not all(isinstance(tag, str) for tag in value):
            raise ValueError("all tag values must be strings")
        self._tags = value

    def to_dict(self):
        return {
            "name": self.name,
            "base_route": self.base_route,
            "description": self.description,
            "tags": self.tags
        }


class AuthenticationConfig:
    """Object representation of the Authentication config section of a Lowball config.

    Defines the authentication settings for token generation.

    :param default_token_life: If no token life is requested when generating a token, this value
        is used
    :type default_token_life: int
    :param max_token_life: the maximum number of seconds that a token can be valid for
    :type max_token_life: int
    :param token_secret: The secret that is used to encode JWT tokens for authentication
    :type token_secret: str
    """

    def __init__(self, default_token_life=3600, max_token_life=2592000, token_secret="CHANGE_ME", **kwargs):
        self.default_token_life = default_token_life
        self.max_token_life = max_token_life
        self.token_secret = token_secret

    @property
    def default_token_life(self):
        return self._default_token_life

    @default_token_life.setter
    def default_token_life(self, value):
        if isinstance(value, str):
            try:
                value = int(value)
            except ValueError:
                pass
        if not isinstance(value, int) or isinstance(value, bool):
            raise TypeError("default_token_life must be an integer")
        self._default_token_life = value

    @property
    def max_token_life(self):
        return self._max_token_life

    @max_token_life.setter
    def max_token_life(self, value):
        if isinstance(value, str):
            try:
                value = int(value)
            except ValueError:
                pass
        if not isinstance(value, int) or isinstance(value, bool):
            raise TypeError("max_token_life must be an integer")
        self._max_token_life = value

    @property
    def token_secret(self):
        return self._token_secret

    @token_secret.setter
    def token_secret(self, value):
        if not isinstance(value, str):
            raise TypeError("token secret must be a string")
        self._token_secret = value

    def to_dict(self):
        return {
            "default_token_life": self.default_token_life,
            "max_token_life": self.max_token_life,
            "token_secret": self.token_secret,
        }


def config_from_object(config_object):
    """Generate a Config object from a dictionary object.

    :type config_object: dict
    :param config_object: dictionary representation of a config object
    :rtype: Config
    :return: Object representation of a  Lowball config
    """
    if not isinstance(config_object, dict):
        raise TypeError("config object must be a dictionary")

    meta = MetaConfig(**config_object.get("meta", {}))
    authentication = AuthenticationConfig(**config_object.get("authentication", {}))
    application = config_object.get("application")
    auth_provider = config_object.get("auth_provider")
    auth_db = config_object.get("auth_db")
    logging = config_object.get("logging")

    return Config(meta=meta, authentication=authentication, application=application, auth_provider=auth_provider,
                  auth_db=auth_db, logging=logging)


def config_from_file(filepath):
    """Generate a Config object from the data in a config file.

    Both YAML (.yaml, .yml) and JSON (.json) config files are supported by this
    method. This method will convert the data in either type of file to a dictionary
    and then pass that data to the :func:`config_from_object` method. See documentation
    for that method for information about required data.

    :type filepath: str, :class:`pathlib.Path`
    :param filepath: path to the config file on the system
    :rtype: Config
    :return: Object representation of a  Lowball config
    """
    if not isinstance(filepath, (str, Path)):
        raise TypeError("filepath must be a string or a pathlib.Path object")

    with open(filepath, "r") as config_file:
        config_data = config_file.read()

    try:
        yaml = YAML()
        config_object = dict(yaml.load(config_data))
    except:
        raise ValueError("invalid config format")

    return config_from_object(config_object)


class Config:
    """Object representation of a Lowball config.

    :param meta: Metadata about the Lowball application
    :type meta: MetaConfig
    :param authentication: Data needed for authentication operations
    :type authentication: AuthenticationConfig
    :param application: config data contained in the `application` field of the config
    :type application: dict, optional
    :param auth_provider: config data contained in the `auth_provider` field of the config
    :type auth_provider: dict, optional
    :param auth_db: config data contained in the `auth_db` field of the config
    :type auth_db: dict, optional
    :param logging: config data contained in the `logging` field of the config
    :param kwargs: Used to catch any values that are mapped into the config object that
        are not supported fields. The idea here is to make it very easy to use `Config(**config)`
        without breaking everything. These values are not used or stored.
    :type kwargs: dict
    """

    def __init__(self, meta=MetaConfig(), authentication=AuthenticationConfig(), application=None, auth_provider=None, auth_db=None, logging=None,
                 **kwargs):
        self.meta = meta
        self.authentication = authentication
        self.application = application if application is not None else {}
        self.auth_provider = auth_provider if auth_provider is not None else {}
        self.auth_db = auth_db if auth_db is not None else {}
        self.logging = logging if logging is not None else {}

    @property
    def meta(self):
        return self._meta

    @meta.setter
    def meta(self, value):
        if not isinstance(value, MetaConfig):
            raise TypeError("meta must be a MetaConfig object")
        self._meta = value

    @property
    def authentication(self):
        return self._authentication

    @authentication.setter
    def authentication(self, value):
        if not isinstance(value, AuthenticationConfig):
            raise TypeError("authentication must be a AuthenticationConfig object")
        self._authentication = value

    @property
    def application(self):
        return self._application

    @application.setter
    def application(self, value):
        if not isinstance(value, dict):
            raise TypeError("application must be a dictionary")
        self._application = value

    @property
    def auth_provider(self):
        return self._auth_provider

    @auth_provider.setter
    def auth_provider(self, value):
        if not isinstance(value, dict):
            raise TypeError("auth_provider must be a dictionary")
        self._auth_provider = value

    @property
    def auth_db(self):
        return self._auth_db

    @auth_db.setter
    def auth_db(self, value):
        if not isinstance(value, dict):
            raise TypeError("auth_db must be a dictionary")
        self._auth_db = value

    @property
    def logging(self):
        return self._logging

    @logging.setter
    def logging(self, value):
        if not isinstance(value, dict):
            raise TypeError("logging must be a dictionary")
        self._logging = value

    def to_dict(self):
        """Return a dictionary representation of the Config object."""
        return {
            "meta": self.meta.to_dict(),
            "authentication": self.authentication.to_dict(),
            "application": self.application,
            "auth_provider": self.auth_provider,
            "auth_db": self.auth_db,
            "logging": self.logging
        }


__all__ = [
    "Config",
    "MetaConfig",
    "config_from_object",
    "config_from_file",
    "AuthenticationConfig"
]
